src/shims/unix/mem.rs

```diff
@@ -0,0 +1,214 @@
+                "mmap(addr: {addr}, length: {length}, prot: {prot:x}, flags: {flags:0x}, fd: {fd}, offset: {offset})"
+            );
+        };
+
+        let prot_read = this.eval_libc_i32("PROT_READ")?;
+        let prot_write = this.eval_libc_i32("PROT_WRITE")?;
+        let map_private = this.eval_libc_i32("MAP_PRIVATE")?;
+        let map_anonymous = this.eval_libc_i32("MAP_ANONYMOUS")?;
+
+        // Only one of MAP_PRIVATE, MAP_SHARED, or MAP_SHARED_VALIDATE may be passed
+        if flags & this.eval_libc_i32("MAP_PRIVATE")? == 0 {
+            throw_unsup_format!("Miri does not support MAP_SHARED or MAP_SHARED_VALIDATE");
+        }
+
+        if flags & this.eval_libc_i32("MAP_STACK")? > 0 {
+            throw_unsup_format!("Miri does not support MAP_STACK");
+        }
+
+        if prot & this.eval_libc_i32("PROT_EXEC")? > 0 {
+            print_args();
+            throw_unsup_format!("Miri does not support mapping executable pages");
+        }
+
+        if offset != 0 {
+            print_args();
+            throw_unsup_format!("Miri does not support non-zero offsets to mmap (yet)");
+        }
+
+        if !this.ptr_is_null(addr)? {
+            print_args();
+            throw_unsup_format!("Miri does not support non-null pointers to mmap");
+        }
+
+        if length == 0 {
+            print_args();
+            this.set_last_error(Scalar::from_i32(this.eval_libc_i32("EINVAL")?))?;
+            return Ok(Pointer::null());
+        }
+
+        let align = Align::from_bytes(PAGE_SIZE).unwrap();
+        let map_length = ((length + PAGE_SIZE - 1) / PAGE_SIZE) * PAGE_SIZE;
+
+        if (flags == map_private | map_anonymous) || (flags == map_private && fd != -1) {
+            // mmap as a memory allocator
+            let ptr = this.allocate_ptr(
+                Size::from_bytes(map_length),
+                align,
+                MiriMemoryKind::Mmap.into(),
+            )?;
+            // We just allocated this, the access is definitely in-bounds and fits into our address space.
+            // mmap guarantees new mappings are zero-init.
+            this.write_bytes_ptr(
+                ptr.into(),
+                std::iter::repeat(0u8).take(usize::try_from(map_length).unwrap()),
+            )
+            .unwrap();
+            let (prov, offset) = ptr.into_parts();
+            let ptr = Pointer::new(Some(prov), offset);
+
+            this.machine.mappings.push(Mapping {
+                ptr,
+                alloc_id: match prov {
+                    Provenance::Concrete { alloc_id, .. } => alloc_id,
+                    Provenance::Wildcard =>
+                        unreachable!("allocate_ptr should not return a Wildcard pointer"),
+                },
+                len: map_length,
+                can_read: prot & prot_read > 0,
+                can_write: prot & prot_write > 0,
+            });
+
+            // If we were passed an fd, populate the first length bytes from the file
```

Comment:
I don't know much about mmap, I just assumed that it meant that changing the file would mean we pick up the changes in the mmapped memory, too. So copying the memory would be wrong.

Or at the very least, mmapping the same file twice would give us racy shared memory or sth

Reply:
`MAP_PRIVATE` causes any changes to the mapped memory range to not affect the backing file, but kept in memory. Changes to the backing file can affect the mapped memory range though, but in most cases that is not what the user wants anyway. There is just no api for mmaping a virtual copy of a file.

Reply:
can/should we poison the file handle at least to prevent it from being used to write to the file?

Reply:
I don't think that would be accurately reproducing the API. But this is good feedback, I think the thing to do here is remove the file mapping support entirely until I figure out how to do MMIO, then answer questions like this when propagating changes to the mapping is on the table.

Reply:
Or leave it in, make it emit a warning about an incomplete feature and open an issue.